indexing-service/src/main/java/io/druid/indexing/common/actions/TaskActionToolbox.java

```diff
@@ -110,4 +114,30 @@ public boolean apply(TaskLock taskLock)
 
     return true;
   }
+
+  // Verify that the segment intervals are covered by lock which is upgraded
+  public boolean locksAreUpgraded (Task task, Set<DataSegment> segments) {
+    // Get the locks from TaskLockbox (at overlord)
+    final List<TaskLock> taskLocks = getTaskLockbox().findLocksForTask(task);
+    // Each segment should be covered by at least a lock which is upgraded
+    // Assumption here is that check for lock presence has already been done
+    // as this check will succeed if there is no lock match for given interval and datasource
+    for (final DataSegment segment : segments) {
+      final boolean isUpgraded = Iterables.all(
+          taskLocks, new Predicate<TaskLock>()
+          {
+            @Override
+            public boolean apply(TaskLock input)
+            {
+              return !(input.getInterval().contains(segment.getInterval())
+                     && input.getDataSource().equals(segment.getDataSource())) || input.isUpgraded();
+            }
+          }
+      );
+      if (!isUpgraded) {
+        return false;
+      }
+    }
+    return true;
```

Comment:
is it possible to merge locksAreUpgraded(..) and taskLockCoversSegment(..) into one method?